test/s3select_test.cpp

```diff
@@ -1878,6 +1878,246 @@ TEST(TestS3selectFunctions, castdatediff)
+    s3selectEngine::csv_object s3_csv_object(&s3select_syntax);
+    std::string s3select_result;
+    std::string input;
+    size_t size = 1;
+    generate_csv(input, size);
+    status = s3_csv_object.run_s3select_on_object(s3select_result, input.c_str(), input.size(), 
+        false, // dont skip first line 
+        false, // dont skip last line
+        true   // aggregate call
+        ); 
+    ASSERT_EQ(status, 0); 
+    ASSERT_EQ(s3select_result, std::string("foobar,\n"));
+}
+
+TEST(TestS3selectFunctions, trim7)
+{
+    s3select s3select_syntax;
+    const std::string input_query = "select substr(trim(both from '   foobar   '),2,3) from stdin ;" ;
+    auto status = s3select_syntax.parse_query(input_query.c_str());
+    ASSERT_EQ(status, 0);
+    s3selectEngine::csv_object s3_csv_object(&s3select_syntax);
+    std::string s3select_result;
+    std::string input;
+    size_t size = 1;
+    generate_csv(input, size);
+    status = s3_csv_object.run_s3select_on_object(s3select_result, input.c_str(), input.size(), 
+        false, // dont skip first line 
+        false, // dont skip last line
+        true   // aggregate call
+        ); 
+    ASSERT_EQ(status, 0); 
+    ASSERT_EQ(s3select_result, std::string("oob,\n"));
+} 
+
+TEST(TestS3selectFunctions, trim8)
+{
+    s3select s3select_syntax;
+    const std::string input_query = "select substr(trim(both from '12'  '1112211foobar22211122'),1,6) from stdin ;" ;
+    auto status = s3select_syntax.parse_query(input_query.c_str());
+    ASSERT_EQ(status, 0);
+    s3selectEngine::csv_object s3_csv_object(&s3select_syntax);
+    std::string s3select_result;
+    std::string input;
+    size_t size = 1;
+    generate_csv(input, size);
+    status = s3_csv_object.run_s3select_on_object(s3select_result, input.c_str(), input.size(), 
+        false, // dont skip first line 
+        false, // dont skip last line
+        true   // aggregate call
+        ); 
+    ASSERT_EQ(status, 0); 
+    ASSERT_EQ(s3select_result, std::string("foobar,\n"));
+} 
+
+TEST(TestS3selectFunctions, trim9)
+{
+    s3select s3select_syntax;
+    const std::string input_query = "select cast(trim(both from \"12\"  \"111221134567822211122\") as int) + 5 from stdin ;" ;
+    auto status = s3select_syntax.parse_query(input_query.c_str());
+    ASSERT_EQ(status, 0);
+    s3selectEngine::csv_object s3_csv_object(&s3select_syntax);
+    std::string s3select_result;
+    std::string input;
+    size_t size = 1;
+    generate_csv(input, size);
+    status = s3_csv_object.run_s3select_on_object(s3select_result, input.c_str(), input.size(), 
+        false, // dont skip first line 
+        false, // dont skip last line
+        true   // aggregate call
+        ); 
+    ASSERT_EQ(status, 0); 
+    ASSERT_EQ(s3select_result, std::string("345683,\n"));
+}
+
+TEST(TestS3selectFunctions, trim10)
+{
+    s3select s3select_syntax;
+    const std::string input_query = "select trim(trim(leading from \"   foobar   \")) from stdin ;" ;
```

Comment:
good test. it verify the trim on trim create correct AST.

its even better, for one trim to operate on one side with specific chars, and other trim to remove the opposite side with different chars.